### PR TITLE
Add NSNet2 sample

### PR DESCRIPTION
**Pull request by huningxin**
Implement the NSNet2 for noise suppression. NSNet2 is one model of [first-wave models](https://github.com/webmachinelearning/webnn/blob/master/op_compatibility/first_wave_models.md) and used as example for key scenarios in [explainer](https://github.com/webmachinelearning/webnn/blob/master/explainer.md#key-scenarios).

Preview: https://huningxin.github.io/webnn-samples/nsnet2/

Test audios (from [rnnoise demo](https://jmvalin.ca/demo/rnnoise/)):
 * [babble](https://jmvalin.ca/demo/rnnoise/samples/babble_15dB.opus)
 * [car](https://jmvalin.ca/demo/rnnoise/samples/car_15dB.opus)
 * [street](https://jmvalin.ca/demo/rnnoise/samples/street_15dB.opus)

@anssiko @wchao1115 @pyu10055 @BruceDai , please take a look. Thanks!

### CODE REVIEW

**Review by anssiko — APPROVED** [review]
Thanks! As a future enhancement (new issue), perhaps allow the selection of those test audio clips from the UI, e. g. from a drop down similar to the code editor.

**Review by BruceDai — APPROVED** [review]
Thanks @huningxin. 

I ran this NSNet2 sample with following Wav / Opus audio files

1.	With a Wav audio file (Size  4.37MB/Length 00:00:23/Bit rate 1536 kbps), sample Passed, and the Denoised Audio file would be a Wav file of 747KB Size with 256kbps Bit rate. 
2.	With two Opus audio files
a.	First file (Size 2.25MB/Length 00:02:54/Audio sample rate 48.000kHz), Chrome crashed, Was its’ length too long?   
b.	I cut first file to be Size 391KB/Length 00:00:30 (Audio sample rate 48.000kHz), sample Passed, and the Denoised Audio file would be a Wav file of 937KB Size with 256kbps Bit rate.

And on UI, I think it would be friendly to user if clearly informed user to choose Wav or Opus audio files, thanks.

**Comment by huningxin** [comment]
Thanks @anssiko and @BruceDai !

> As a future enhancement (new issue), perhaps allow the selection of those test audio clips from the UI, e. g. from a drop down similar to the code editor.

Good point. I'll try to implement that in this PR and let you review again.

> 2\. First file (Size 2.25MB/Length 00:02:54/Audio sample rate 48.000kHz), Chrome crashed, Was its’ length too long?

It might be caused by GPU memory issues. I usually tested audio files less than 1 min. I'll try to look into this issue.


> And on UI, I think it would be friendly to user if clearly informed user to choose Wav or Opus audio files

Good point. As the UI says "choose an audio file", probably it can accept more audio types, e.g. .mp3 and .ogg etc.,

**Comment by anssiko** [comment]
Thanks @huningxin! To be exact, I wanted to say here:

>allow the selection of those test audio clips

allow the selection of those test audio clips _in addition to the current file picker_.

The current UI is very functional, but this enhancement idea would make it work on mobile browsers that may not provide access to the local filesystem.

**Comment by huningxin** [comment]
@anssiko , 

> allow the selection of those test audio clips _in addition to the current file picker_.

I implemented this feature in the new commits. Please take a look at the updated [preview](https://huningxin.github.io/webnn-samples/nsnet2/).

The screenshot is

image

The top three are sample audios. The last one is to launch the file picker.

**Comment by huningxin** [comment]
@BruceDai 

> > 2. First file (Size 2.25MB/Length 00:02:54/Audio sample rate 48.000kHz), Chrome crashed, Was its’ length too long?
> 
> It might be caused by GPU memory issues. I usually tested audio files less than 1 min. I'll try to look into this issue.

I fixed the memory leak issues of this example in the new commits. With the fix, after each denoising, the `numTensors` of tf.js tensors are kept as 18 (for the constants of NSNet2). However, I found the `numBytesInGPU` of tf.js keeps increasing after each denoising till Chrome's GPU process gets killed.

It seems to me that the example code and webnn-polyfill successfully release (dispose) all immediate tensors, but the the GPU memory used by tf.js still increases. Is this a memory leak issue within tf.js? @pyu10055 , do you have any thoughts?

You can reproduce this by opening the JavaScript console and type `_tfengine.memory()` which will tell how many tensors and GPU memory used by tf.js.

For my case, when NSNet2 is ready, before process any files, the `_tfengine.memory()` output is
```
numBytes: 10879044
numBytesInGPU: 11150096
numBytesInGPUAllocated: 79964368
numBytesInGPUFree: 68814272
numDataBuffers: 18
numTensors: 18
unreliable: false
```

After processing "Babble noise" audio, the output of `_tfengine.memory()` is
```
numBytes: 10879044
numBytesInGPU: 38198096
numBytesInGPUAllocated: 121521624
numBytesInGPUFree: 83323528
numDataBuffers: 18
numTensors: 18
unreliable: false
```

The `numTensors` is same as 18, however, the `numBytesInGPU` gets increased `38198096` vs. `11150096`.

@BruceDai , probably you can help verify on your side.

**Comment by anssiko** [comment]
@huningxin now the UI is perfect, thanks!

This is a compelling demonstration of the [noise suppression use case](https://webmachinelearning.github.io/webnn/#usecase-noise-suppression) we have in the spec. Feel free to add NSNet2 as an additional reference for that use case into the spec, as well as add a link to this sample from the explainer [key scenarios](https://github.com/webmachinelearning/webnn/blob/master/explainer.md#key-scenarios) section.

File: nsnet2/nsnet2.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
+
+'use strict';
+
+function sizeOfShape(shape) {
+  return shape.reduce((a, b) => {
+    return a * b;
+  });
+}
+
+export class NSNet2 {
```
Comment (huningxin) [inline nsnet2/nsnet2.js:10]: @wchao1115, could you please help review this NSNet2 implementation? If the code looks good, I could submit a patch to update the NSNet2 example code in the [explainer](https://github.com/webmachinelearning/webnn/blob/master/explainer.md#key-scenarios) according to this one. 

What do you think?

**Comment by huningxin** [comment]
Ping @pyu10055 for the memory leaks issue of tf.js, please find the details at https://github.com/webmachinelearning/webnn-samples/pull/22#issuecomment-730217119

**Comment by huningxin** [comment]
@miaobin, I added a commit to load weights from npy files. It might be useful for your rnnoise example. Please take a look. Thanks!

**Comment by huningxin** [comment]
> Ping @pyu10055 for the memory leaks issue of tf.js, please find the details at [#22 (comment)](https://github.com/webmachinelearning/webnn-samples/pull/22#issuecomment-730217119)

reported an issue to tf.js https://github.com/tensorflow/tfjs/issues/4378

File: nsnet2/nsnet2.js (hunk 1 above)
Comment (wchao1115) [reply nsnet2/nsnet2.js:10]: looks great! Thanks!

**Comment by huningxin** [comment]
Updated this PR with tf.js 2.8.5. The memory leak issue is verified to be fixed.

Also depends on webnn-polyfill tf.js 2.8.5 upgrading https://github.com/webmachinelearning/webnn-polyfill/pull/33.

**Review by pyu10055 — APPROVED** [review]
The demo sounds great, awesome work!

**Comment by huningxin** [comment]
Thanks for the review! I am going to merge this PR and deploy to github.io. Once deployed, I'll send a PR to webnn explainer to add the link.

**Pull request merged**